city_scrapers/spiders/il_lottery.py

```diff
@@ -0,0 +1,135 @@
+                start=self._parse_start(item),
+                end=None,
+                all_day=False,
+                time_notes='',
+                location={
+                    'name': 'Chicago Lottery Office',
+                    'address': '122 South Michigan Avenue, 19th Floor, Chicago, IL 60603'
+                },
+                source=response.url,
+            )
+            meeting['id'] = self._get_id(meeting)
+            meeting['status'] = self._get_status(meeting, text=item)
+            meeting['links'] = self._parse_agenda_links(response, date=meeting['start'].date())
+
+            yield meeting
+
+    def _parse_title(self, item):
+        """Parse or generate meeting title."""
+        suffix = ''
+        if 'QTR' in item:
+            suffix = " Quarterly Meeting"
+        elif 'Special' in item:
+            suffix = ' Special Meeting'
+        return "Lottery Control Board{}".format(suffix)
+
+    @staticmethod
+    def parse_time(source):
+        """Returns times given string with format h:mm am|pm"""
+        time_regex = re.compile(r'([1-9]):([0-5][0-9])\s?(am|pm)')
+        hour, minute, period = time_regex.search(source).groups()
+
+        hour = int(hour)
+        minute = int(minute)
+        if (period == 'pm') and (hour != 12):
+            hour += 12
+        return time(hour, minute)
+
+    @staticmethod
+    def parse_day(source):
+        """Returns date"""
+        # search for dates with '/' (ex: 08/16/19)
+        if "/" in source:
+            m = re.search(
+                r'.+[,-]\s(?P<month>\d{2})\/(?P<day>\d{2})\/(?P<year>\d{2})\s.+', source.strip()
+            )
+        # search for date in format "[month] [dd], [yyyy]" (ex: 'May 15, 2019')
+        else:
+            m = re.search(
+                r'.+[,-]\s*(?P<month>[A-Za-z]+)\s(?P<day>\d{1,2}),\s(?P<year>\d{4}).*',
+                source.strip()
+            )
+
+        dt = dateutil.parser.parse(
+            '{mo} {day} {year}'.format(
+                mo=m.group('month'), day=m.group('day'), year=m.group('year')
+            )
+        )
+        return dt.date()
+
+    def _parse_start(self, item):
+        """Parse start date and time."""
+        return datetime.combine(self.parse_day(item), self.parse_time(item))
+
+    def _parse_upcoming(self, response):
+        """Returns a list of lines with dates following the string 'Upcoming Meeting Dates' """
+
+        meeting_xpath = '//p[contains(., "Upcoming meeting dates")]'
+
+        # future meetings are separated by <br> tags
+        meeting_lines = response.xpath(meeting_xpath).css('p').get().split("<br>")
+
+        # only keep <br> lines that include a weekday
+        weekdays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
+        meeting_lines = [x for x in meeting_lines if any(weekday in x for weekday in weekdays)]
+
+        return meeting_lines
+
+    def _parse_agenda_links(self, response, date):
+        '''
```

Comment:
This is a good approach, but it looks like it's re-parsing the full response for every meeting. It might be a little more straightforward to run this before iterating through the meetings and return a dictionary of date keys and link list values, so that in the for loop you can just have something like:

```python
meeting['links'] = date_link_map.get(meeting['start'].date(), [])
```

Reply:
Thanks for the suggestion! I created two dictionaries, agenda_links and minutes_links, that store date keys and link values. Those dictionaries are created before iterating through the meetings.

Reply:
This update looks good! It looks like it's missing some cases where there are only minutes but no agenda though, like October 10, 2012. You should be able to combine the agenda and minutes dictionaries using a `defaultdict` of lists of links, similar to what we do in [`chi_il_medical_district`](https://github.com/City-Bureau/city-scrapers/blob/42fab8ff5a9763a29f9192ef2cdda31228f74970/city_scrapers/spiders/chi_il_medical_district.py#L108), that way you only have to iterate through once, checking if "Agenda" or "Minutes" is in the link and then just pull links from the `defaultdict` with `meeting['links'] = date_link_map[meeting['start'].date()]` to make it even simpler. Let me know if that example isn't clear!

Reply:
Thanks for the tip! I implemented the `defaultdict` in the [latest commit](https://github.com/City-Bureau/city-scrapers/pull/820/commits/b7e82f758bc417cdf32445ec5c468789df8694ee).